plutus-use-cases/src/Language/Plutus/Coordination/Contracts/CrowdFunding.hs

```diff
@@ -0,0 +1,150 @@
+
+            infixr 3 &&
+            (&&) :: Bool -> Bool -> Bool
+            (&&) = Prim.error
+
+            infixr 3 ||
+            (||) :: Bool -> Bool -> Bool
+            (||) = Prim.error
+
+            -- | Check that a pending transcation is signed by the private key of the given
+            --   public key.
+            signedByT :: PendingTx -> PubKey -> Bool
+            signedByT = Prim.error
+
+            maybe :: b -> (a -> b) -> Maybe a -> b
+            maybe = Prim.error
+
+            PendingTx pendingTxBlockHeight _ pendingTxTransaction = p
+
+            Tx (t1, t2) _ = pendingTxTransaction
+            TxIn (TxOutRef v1 _ _) _ = t1
+            TxIn (TxOutRef v2 _ _) _ = t2
+
+
+            pledgedFunds = v1 + v2
+            -- Check that a refund transaction only spends the amount that was
+            -- pledged by the contributor identified by `contribPubKey`
+            contributorOnly = maybe False (\k ->
+                                signedBy t1 k && signedBy t2 k) contribPubKey
+            payToOwner   = pendingTxBlockHeight > campaignDeadline &&
+                           pendingTxBlockHeight <= campaignCollectionDeadline &&
+                           pledgedFunds >= campaignTarget &&
+                           signedByT p campaignOwner
+            -- In case of a refund, we can only collect the funds that
+            -- were committed by this contributor
+            refundable   = pendingTxBlockHeight > campaignCollectionDeadline &&
+                           contributorOnly &&
+                           maybe False (signedByT p) contribPubKey
+        in
+        if payToOwner || refundable then () else Prim.error)
+
+-- | Given the campaign data and the output from the contributing transaction,
+--   make a trigger that fires when the transaction can be refunded.
+refundTrigger :: Campaign -> Address -> EventTrigger
+refundTrigger Campaign{..} t = And
+    (FundsAtAddress [t]  (GEQ 1))
+    (BlockHeightRange (GEQ $ succ campaignCollectionDeadline))
+
+-- | Given the public key of the campaign owner, generate an event trigger that
+-- fires when the funds can be collected.
+collectFundsTrigger :: Campaign -> [Address] -> EventTrigger
+collectFundsTrigger Campaign{..} ts = And
+    (FundsAtAddress ts $ GEQ campaignTarget)
+    (BlockHeightRange $ Interval campaignDeadline campaignCollectionDeadline)
+
+refund :: TxOutRef -> TxM [TxOutRef]
+refund ref = do
+    kp <- lookupMyKeyPair
+    let i = txInSign ref kp
+        o = TxOutPubKey value (pubKey kp)
+    submitTransaction $ Tx {
+      txInputs = (i, i),
+      txOutputs = (o, o)
+    } where
+      value = txOutRefValue ref - standardTxFee -- TODO: Fee should be inserted by wallet
+
+-- | Collect all campaign funds (campaign owner)
+--
+-- NB: Simplifing assumption: the number of contributions doesn’t
+--     exceed the number of inputs that we can put into a single
+--     transaction.
+--
+collect :: (TxOutRef, TxOutRef) -> TxM [TxOutRef]
```

Comment:
Presumably there also has to be some way to find all the refs that you can spend? Here we know there are just two.